cirq/_compat.py

```diff
@@ -83,3 +85,64 @@ def decorated_func(*args, **kwargs) -> Any:
         return decorated_func
 
     return decorator
+
+
+def deprecated_parameter(
+        *,
+        deadline: str,
+        fix: str,
+        func_name: Optional[str] = None,
+        parameter_desc: str,
+        match: Callable[[Tuple[Any, ...], Dict[str, Any]], bool],
+        rewrite: Optional[
+            Callable[[Tuple[Any, ...], Dict[str, Any]],
+                     Tuple[Tuple[Any, ...], Dict[str, Any]]]] = None,
+) -> Callable[[Callable], Callable]:
+    """Marks a function parameter as deprecated.
+
+    Also handles rewriting the deprecated parameter into the new signature.
+
+    Args:
+        deadline: The version where the parameter will be deleted (e.g. "v0.7").
+        fix: A complete sentence describing what the user should be using
+            instead of this particular function (e.g. "Use cos instead.")
+        func_name: How to refer to the function.
+            Defaults to `func.__qualname__`.
+        parameter_desc: The name and type of the parameter being deprecated,
+            e.g. "janky_count" or "janky_count keyword" or
+            "positional janky_count".
+        match: Detects if the deprecated parameter is present in the
+            args/kwargs.
```

Comment:
I think this could be explained better.  I had to go to the tests to understand how this is supposed to work.  I totally didn't understand that the match parameter was the triggering condition for the parameter.  Maybe add some examples?

Reply:
Done